### PR TITLE
feat: Memory limited merge join

### LINKED ISSUE #5220: Use `MemoryManager` in Join operators, and return errors if the memory budget is exceeded
See description on https://github.com/apache/arrow-datafusion/issues/3941

The idea is to apply the same treatment to Join operators that was done to GroupBy in https://github.com/apache/arrow-datafusion/issues/3940 and throw an error if their memory budget is exceeded rather than unlimited allocating.

All the operators in https://github.com/apache/arrow-datafusion/tree/master/datafusion/core/src/physical_plan/joins need to be evaluated. 

Items to do: 
- [x] Apply limits to NLJ https://github.com/apache/arrow-datafusion/issues/5162
- [x] Apply limits to Cross Join
- [x] Apply limits to Hash Join
- [ ] Apply limits to Sort Merge Join

cc @DDtKey  and @crepererum who have expressed interest in this project in the past

### PR DESCRIPTION
**Pull request by korowa**
# Which issue does this PR close?

Closes #5220.

# Rationale for this change

Control over memory allocation in `SortMergeJoinExec`

# What changes are included in this PR?

- `SMJStream.reservation` -- memory reservation to control buffered data size -- it may require significant amount of memory, for example, in case of multiple batches with the same join key values
- `peak_memory_used` -- during join execution, normally, both `try_grow` and `shrink` methods are called  -- on adding new buffered batches and on flushing them if join key has changed, respectively. In this case it seems reasonable to track only peak memory for each partition -- after that it's summed up across all partitions which is not really precise, but still valuable in terms of showing "worst case" (in fact peak memory used, probably, can be lower)
- memory limit tests now are able to run with configurable `SessionConfig` and perform multiple checks on returned error (used to ensure that memory allocation failed in specific operator/stream)

# Are these changes tested?

- existsing test cases (shrinks don't lead to negative size)
- overallocation tests for all join types (if accepted by merge join operator)
- SQL-level test for overallocation

# Are there any user-facing changes?

`SortMergeJoinExec` should now respect runtime memory limitations.

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thank you @korowa  -- this looks great. I reviewed the code and made sure the reservation was updated when the buffered batch was pushed/poped.

cc @richox  who I believe contributed the original implementation in https://github.com/apache/arrow-datafusion/pull/2242

File: datafusion/core/src/physical_plan/joins/sort_merge_join.rs [hunk 1]
```diff
@@ -2212,4 +2258,135 @@ mod tests {
...
+                sort_options.clone(),
+                false,
+            )?;
+
+            let stream = join.execute(0, task_ctx)?;
+            let err = common::collect(stream).await.unwrap_err();
+
+            assert_contains!(
+                err.to_string(),
+                "Resources exhausted: Failed to allocate additional"
```
Comment (alamb) [inline datafusion/core/src/physical_plan/joins/sort_merge_join.rs:2309]: Thank you for the unit test

File: datafusion/core/src/physical_plan/joins/sort_merge_join.rs [hunk 2]
```diff
@@ -505,15 +516,33 @@ struct BufferedBatch {
...
+        // + worst case null_joined (as vector capacity)
+        // + Range size (16)
+        // + size of this estimation (8)
+        let size_estimation = batch.get_array_memory_size()
+            + join_arrays
+                .iter()
+                .map(|arr| arr.get_array_memory_size())
+                .sum::<usize>()
+            + batch.num_rows().next_power_of_two() * 8
+            + 24;
```
Comment (Dandandan) [inline datafusion/core/src/physical_plan/joins/sort_merge_join.rs:538]:
```suggestion
            + sizeof::<Range>()
            + sizeof::<usize>();
```

Comment (korowa) [reply datafusion/core/src/physical_plan/joins/sort_merge_join.rs]: True, it may vary. I've fixed it in separate commit, thank you!

**Pull request merged**